src/Buffer.mo

```diff
@@ -1,166 +1,1751 @@
+  ///
+  /// Runtime: O(size)
+  ///
+  /// Space: O(size)
+  public func fromVarArray<X>(array : [var X]) : Buffer<X> {
+    let newBuffer = Buffer<X>(newCapacity(array.size()));
+
+    for (element in array.vals()) {
+      newBuffer.add(element);
+    };
+
+    newBuffer;
+  };
+
+  /// Creates a buffer containing elements from `iter`.
+  ///
+  /// Runtime: O(size)
+  ///
+  /// Space: O(size)
+  public func fromIter<X>(iter : { next : () -> ?X }) : Buffer<X> {
+    let newBuffer = Buffer<X>(DEFAULT_CAPACITY); // can't get size from `iter`
+
+    for (element in iter) {
+      newBuffer.add(element);
+    };
+
+    newBuffer;
+  };
+
+  /// Reallocates the array underlying `buffer` such that capacity == size.
+  ///
+  /// Runtime: O(size)
+  ///
+  /// Space: O(size)
+  public func trimToSize<X>(buffer : Buffer<X>) {
+    let size = buffer.size();
+    if (size < buffer.capacity()) {
+      buffer.reserve(size);
+    };
+  };
+
+  /// Creates a new buffer by applying `f` to each element in `buffer`.
+  ///
+  /// Runtime: O(size)
+  ///
+  /// Space: O(size)
+  ///
+  /// *Runtime and space assumes that `f` runs in O(1) time and space.
+  public func map<X, Y>(buffer : Buffer<X>, f : X -> Y) : Buffer<Y> {
+    let newBuffer = Buffer<Y>(buffer.capacity());
+
+    for (element in buffer.vals()) {
+      newBuffer.add(f element);
+    };
+
+    newBuffer;
+  };
+
+  /// Applies `f` to each element in `buffer`.
+  ///
+  /// Runtime: O(size)
+  ///
+  /// Space: O(size)
+  ///
+  /// *Runtime and space assumes that `f` runs in O(1) time and space.
+  public func iterate<X>(buffer : Buffer<X>, f : X -> ()) {
+    for (element in buffer.vals()) {
+      f element;
+    };
+  };
+
+  /// Applies `f` to each element in `buffer` and its index.
+  ///
+  /// Runtime: O(size)
+  ///
+  /// Space: O(size)
+  ///
+  /// *Runtime and space assumes that `f` runs in O(1) time and space.
+  public func mapEntries<X, Y>(buffer : Buffer<X>, f : (Nat, X) -> Y) : Buffer<Y> {
+    let newBuffer = Buffer<Y>(buffer.capacity());
+
+    var i = 0;
+    let size = buffer.size();
+    while (i < size) {
+      newBuffer.add(f(i, buffer.get(i)));
+      i += 1;
+    };
+
+    newBuffer;
+  };
+
+  /// Creates a new buffer by applying `f` to each element in `buffer`,
+  /// and keeping all non-null elements.
+  ///
+  /// Runtime: O(size)
+  ///
+  /// Space: O(size)
+  ///
+  /// *Runtime and space assumes that `f` runs in O(1) time and space.
+  public func mapFilter<X, Y>(buffer : Buffer<X>, f : X -> ?Y) : Buffer<Y> {
+    let newBuffer = Buffer<Y>(buffer.capacity());
+
+    for (element in buffer.vals()) {
+      switch (f element) {
+        case (?element) {
+          newBuffer.add(element);
+        };
+        case _ {};
+      };
+    };
+
+    newBuffer;
+  };
+
+  /// Creates a new buffer by applying `f` to each element in `buffer`.
+  /// If any invocation of `f` produces an #err, returns an #err. Otherwise
+  /// Returns an #ok containing the new buffer.
```

Comment:
```suggestion
  /// If any invocation of `f` produces an `#err`, returns an `#err`. Otherwise
  /// Returns an `#ok` containing the new buffer.
```